pkg/cmd/cli/cmd/env.go

```diff
@@ -349,36 +361,38 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
 	return nil
 }
 
-func updateEnv(existing []kapi.EnvVar, env []kapi.EnvVar, remove []string) []kapi.EnvVar {
+func updateEnv(existing []kapi.EnvVar, env []kapi.EnvVar, remove []string) ([]kapi.EnvVar, bool) {
 	out := []kapi.EnvVar{}
-	covered := sets.NewString(remove...)
+	envUpdated := false
+	toDelete := sets.NewString(remove...)
+
 	for _, e := range existing {
-		if covered.Has(e.Name) {
+		if toDelete.Has(e.Name) {
+			envUpdated = true
 			continue
 		}
-		newer, ok := findEnv(env, e.Name)
+		newer, ok := findEnv(env, e)
 		if ok {
-			covered.Insert(e.Name)
-			out = append(out, newer)
-			continue
+			envUpdated = true
 		}
-		out = append(out, e)
+		out = append(out, newer)
 	}
+
 	for _, e := range env {
-		if covered.Has(e.Name) {
+		if toDelete.Has(e.Name) {
```

Comment:
should be skipping if the env already exists, not if it was in the set to delete... it's an error to delete and set/update the same envvar in the same command (prevented by validation, I'm pretty sure)